exp/runtime/topologymutation/walker.go

```diff
@@ -0,0 +1,220 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topologymutation
+
+import (
+	"context"
+	"encoding/json"
+
+	mergepatch "github.com/evanphx/json-patch/v5"
+	"github.com/pkg/errors"
+	"gomodules.xyz/jsonpatch/v2"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
+	patchvariables "sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches/variables"
+)
+
+// WalkTemplatesOption is some configuration that modifies WalkTemplates behavior.
+type WalkTemplatesOption interface {
+	// ApplyToWalkTemplates applies this configuration to the given WalkTemplatesOptions.
+	ApplyToWalkTemplates(*WalkTemplatesOptions)
+}
+
+// WalkTemplatesOptions contains options for WalkTemplates behavior.
+type WalkTemplatesOptions struct {
+	failForUnknownTypes bool
+	patchFormat         runtimehooksv1.PatchType
+	// TODO: add the possibility to set patchFormat for single patches only, eg. via a func(requestItem) format.
```

Comment:
Q: This would be to have different patch formats for different templates?

Is there a use case where we need this?

Reply:
I also wonder what the up/downside is of using one patch format over another. Are there specific use cases where I would want to pick one vs the other?

(I think that's something that a user of this lib would probably ask themselves as well)

Reply:
AFAIK there is a couple of differences, mostly about how fields of type array are managed; with merge patch, you have to rebuild the entire array, while with json patch you can append or prepend and probably more.

What I was thinking is, as a runtime extension author, if I hit some problem due to the above limitations, I would prefer to change the patch format for one single patch, not for all (containing the blast radius)

Reply:
> AFAIK there is a couple of differences, mostly about how fields of type array are managed; with merge patch, you have to rebuild the entire array, while with json patch you can append or prepend and probably more.

Yup agree.

Just wonder if it's relevant as the patch will be generated automatically anyway and I would hope that both always generate the right patch. And as nobody has to write the patch manually I would assume nobody really cares how the patch exactly looks like as long as applying the patch leads to the expected result.

> What I was thinking is, as a runtime extension author, if I hit some problem due to the above limitations, I would prefer to change the patch format for one single patch, not for all (containing the blast radius)

I'm not sure, but I suspect both formats can achieve the same kind of modifications they just have to include more or less of the patched resources (but this will be done automatically anyway). I think we should explore this further in a follow-up if we think that one of the formats doesn't work correctly in edge cases. Those would be very hard to catch bugs.

Reply:
agreed, it is just a TODO for now